server/src/main/java/org/elasticsearch/action/support/IndicesOptions.java

```diff
@@ -22,138 +24,425 @@
+ * @param wildcardOptions, applies only to wildcard expressions and defines how the wildcards will be expanded and if it will
+ *                        be acceptable to have expressions that results to no indices.
+ * @param generalOptions, applies to all the resolved indices and defines if throttled will be included and if certain type of
+ *                        aliases or indices are allowed, or they will throw an error.
  */
-public record IndicesOptions(EnumSet<Option> options, EnumSet<WildcardStates> expandWildcards) implements ToXContentFragment {
+public record IndicesOptions(ConcreteTargetOptions concreteTargetOptions, WildcardOptions wildcardOptions, GeneralOptions generalOptions)
+    implements
+        ToXContentFragment {
 
-    public enum WildcardStates {
-        OPEN,
-        CLOSED,
-        HIDDEN;
+    public static IndicesOptions.Builder newBuilder() {
+        return new Builder();
+    }
 
-        public static final EnumSet<WildcardStates> NONE = EnumSet.noneOf(WildcardStates.class);
+    public static IndicesOptions.Builder newBuilder(IndicesOptions indicesOptions) {
+        return new Builder(indicesOptions);
+    }
 
-        public static EnumSet<WildcardStates> parseParameter(Object value, EnumSet<WildcardStates> defaultStates) {
-            if (value == null) {
-                return defaultStates;
+    /**
+     * Controls the way the target indices will be handled.
+     * @param allowUnavailableTargets, if false when any of the concrete targets requested does not exist, throw an error
+     */
+    public record ConcreteTargetOptions(boolean allowUnavailableTargets) implements ToXContentFragment {
+        public static final ConcreteTargetOptions ALLOW_UNAVAILABLE_TARGETS = new ConcreteTargetOptions(true);
+        public static final ConcreteTargetOptions ERROR_WHEN_UNAVAILABLE_TARGETS = new ConcreteTargetOptions(false);
+
+        public static ConcreteTargetOptions fromParameter(Object ignoreUnavailableString, ConcreteTargetOptions defaultOption) {
+            if (ignoreUnavailableString == null && defaultOption != null) {
+                return defaultOption;
             }
+            return nodeBooleanValue(ignoreUnavailableString, "ignore_unavailable")
+                ? ALLOW_UNAVAILABLE_TARGETS
+                : ERROR_WHEN_UNAVAILABLE_TARGETS;
+        }
 
-            EnumSet<WildcardStates> states = EnumSet.noneOf(WildcardStates.class);
-            String[] wildcards = nodeStringArrayValue(value);
-            // TODO why do we let patterns like "none,all" or "open,none,closed" get used. The location of 'none' in the array changes the
-            // meaning of the resulting value
-            for (String wildcard : wildcards) {
-                updateSetForValue(states, wildcard);
+        @Override
+        public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+            return builder.field("ignore_unavailable", allowUnavailableTargets);
```

Comment:
Can we pull this out into a `public static String` class variable? I wouldn't want to make this PR change it anywhere it's used, but if we make it public then it encourages people to use it instead of sticking the string `"ignore_unavailable"` everywhere in tests and the like.